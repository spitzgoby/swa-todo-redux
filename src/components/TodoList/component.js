import Button from '@material-ui/core/Button';
import PropTypes from 'prop-types';
import React from 'react';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableFooter from '@material-ui/core/TableFooter';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TodoForm from 'components/TodoForm';
import TodoItem from 'components/TodoItem';
import Typography from '@material-ui/core/Typography';

import './style.scss';

const TodoList = (props) => {

  React.useEffect(() => {
      props.retrieveTodos()
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const getClearTodosButtonProps = () => {
    return {
      className: 'todo-list--clear-todos-button',
      onClick: () => {},
      variant: 'outlined'
    };
  };

  return (
    <div className="todo-list">
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>
              Completed
            </TableCell>
            <TableCell>
              Description
            </TableCell>
            <TableCell>
              Due By
            </TableCell>
            <TableCell />
          </TableRow>
        </TableHead>
        <TableBody>
          {props.todos.map(todoId => <TodoItem key={todoId} todoId={todoId} />)}
        </TableBody>
        <TableFooter>
            { props.user === 'parent' ? <TodoForm /> : null }
        </TableFooter>
      </Table>
      <Typography className="todo-list--completed-count" variant="overline">
        {props.completedTodos.length} of {props.todos.length} completed
      </Typography>
      <Button {...getClearTodosButtonProps()}>
        Clear Completed Todos
      </Button>
    </div>
  );
};

TodoList.propTypes = {
    todos: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
    user: PropTypes.string
};

export default TodoList;
