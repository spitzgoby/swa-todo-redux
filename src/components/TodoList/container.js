import {connect} from 'react-redux';
import {
    clearCompletedTodos,
    getCompletedTodos,
    getTodos,
    retrieveTodos
} from 'modules/Todos'
import {getUser} from 'modules/Users';
import TodoList from './component';

const mapStateToProps = (state) => ({
    completedTodos: getCompletedTodos(state),
    todos: getTodos(state),
    user: getUser(state)
});

const mapDispatchToProps = (dispatch) => ({
    clearCompletedTodos: () => dispatch(clearCompletedTodos()),
    retrieveTodos: () => dispatch(retrieveTodos())
});

export default connect(mapStateToProps, mapDispatchToProps)(TodoList);
