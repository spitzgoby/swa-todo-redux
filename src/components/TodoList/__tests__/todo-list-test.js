import TodoForm from 'components/TodoForm/component';
import TodoList from 'components/TodoList/component';
import { shallow } from 'enzyme';
import { Button, TableBody } from '@material-ui/core';
import React from 'react';

describe('<TodoList />', () => {
    const clearCompletedTodosSpy = jest.fn();

    const mockProps = {
        clearCompletedTodos: clearCompletedTodosSpy,
        completedTodos: [
            {
                completed: true,
                description: 'todo',
                dueBy: '',
                id: 0
            },
        ],
        retrieveTodos: jest.fn(),
        todos: [],
        user: 'parent'
    };
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<TodoList {...mockProps} />);
    });

    it('should render correctly with no todos', () => {
        expect(wrapper).toMatchSnapshot('no-todos');
    });

    it('should render correctly with two todos', () => {
        wrapper = shallow(<TodoList {...mockProps} todos={[1, 2]} />);

        expect(wrapper).toMatchSnapshot('two-todos');
    });

    it('should call retrieveTodos', () => {
        // coming soon!
    });

    it('should render two todos', () => {
        expect(wrapper.find(TableBody).children()).toHaveLength(mockProps.todos.length);
    });

    it('should render TodoForm if user is parent', () => {
        expect(wrapper.find(TodoForm)).toBeDefined();
    });

    it('should render completed count', () => {
        expect(wrapper.find('.todo-list--completed-count').text())
        .toEqual(`${mockProps.completedTodos.length} of ${mockProps.todos.length} completed`);
    });

    it('should clear completed todos', () => {
        wrapper.find(Button).simulate('click', {});

        expect(clearCompletedTodosSpy).toHaveBeenCalled();
    });
})
