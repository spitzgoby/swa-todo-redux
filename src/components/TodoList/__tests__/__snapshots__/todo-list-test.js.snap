// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> should render correctly with no todos: no-todos 1`] = `
<div
  className="todo-list"
>
  <WithStyles(ForwardRef(Table))>
    <WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))>
          Completed
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          Description
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          Due By
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell)) />
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody)) />
    <WithStyles(ForwardRef(TableFooter))>
      <Connect(TodoForm) />
    </WithStyles(ForwardRef(TableFooter))>
  </WithStyles(ForwardRef(Table))>
  <WithStyles(ForwardRef(Typography))
    className="todo-list--completed-count"
    variant="overline"
  >
    1
     of 
    0
     completed
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Button))
    className="todo-list--clear-todos-button"
    onClick={[Function]}
    variant="outlined"
  >
    Clear Completed Todos
  </WithStyles(ForwardRef(Button))>
</div>
`;

exports[`<TodoList /> should render correctly with two todos: two-todos 1`] = `
<div
  className="todo-list"
>
  <WithStyles(ForwardRef(Table))>
    <WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))>
          Completed
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          Description
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          Due By
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell)) />
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody))>
      <Connect(TodoItem)
        key="1"
        todoId={1}
      />
      <Connect(TodoItem)
        key="2"
        todoId={2}
      />
    </WithStyles(ForwardRef(TableBody))>
    <WithStyles(ForwardRef(TableFooter))>
      <Connect(TodoForm) />
    </WithStyles(ForwardRef(TableFooter))>
  </WithStyles(ForwardRef(Table))>
  <WithStyles(ForwardRef(Typography))
    className="todo-list--completed-count"
    variant="overline"
  >
    1
     of 
    2
     completed
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Button))
    className="todo-list--clear-todos-button"
    onClick={[Function]}
    variant="outlined"
  >
    Clear Completed Todos
  </WithStyles(ForwardRef(Button))>
</div>
`;
